kvserver: pass priority at enqueue time to baseQueue.process

wenyihu6 opened the pull request:
Part of: https://github.com/cockroachdb/cockroach/issues/152022
Release note: none 

----

**kvserver: plumb enqueue time priority**

This commit plumbs the enqueue time priority into baseQueue.processReplica,
enabling comparison between the priority at enqueue time and at processing time.
For now, we pass -1 in all cases except when processing replicas directly from
the base queue, where -1 signals that priority verification should be skipped.
No logic change has been made yet to check for priority inversion; future
commits will extend processReplica to validate that processing priority has not
differed significantly from the enqueue time priority.

---

**kvserver: remove priority reset during setProcessing**

Previously, a replicaItem’s priority was cleared when marked as processing, to
indicate it was no longer in the priority queue. This behavior made sense when
the purgatory queue did not track priorities. However, we now need to preserve
priorities for items in purgatory as well since they will be calling into
baseQueue.processReplica. This commit removes the priority reset in
replicaItem.SetProcessing(), ensuring that the enqueue time priority is retained
when replicas are popped from the heap and passed into the purgatory queue
properly. No behavior change should happen from this change.

---

**kvserver: plumb priority at enqueue for purgatory queue**

Previously, replica items in the purgatory queue did not retain their enqueue
time priority. This commit ensures that the priority is preserved so it can be
passed to baseQueue.processReplica when processing items from purgatory.

---

**allocatorimpl: adds a priority assertion to computeAction**

This commit adds an assertion to Allocator.ComputeAction to ensure that priority
is never -1 in cases where it shouldn’t be. Normally, ComputeAction returns
action.Priority(), but we sometimes adjust the priority for specific actions
like AllocatorAddVoter, AllocatorRemoveDeadVoter, and AllocatorRemoveVoter. A
priority of -1 is a special case reserved for processing logic to run even if
there’s a priority inversion. If the priority is not -1, the range may be
re-queued to be processed with the correct priority.

---

**allocatorimpl: add invariants on priority to base queue tests**

This commit adds additional invariants to verify the correctness of priority
plumbing for range items in base queue tests.

wenyihu6 committed:
kvserver: plumb enqueue time priority

This commit plumbs the enqueue time priority into baseQueue.processReplica,
enabling comparison between the priority at enqueue time and at processing time.
For now, we pass -1 in all cases except when processing replicas directly from
the base queue, where -1 signals that priority verification should be skipped.
No logic change has been made yet to check for priority inversion; future
commits will extend processReplica to validate that processing priority has not
differed significantly from the enqueue time priority.

wenyihu6 committed:
kvserver: remove priority reset during setProcessing

Previously, a replicaItem’s priority was cleared when marked as processing, to
indicate it was no longer in the priority queue. This behavior made sense when
the purgatory queue did not track priorities. However, we now need to preserve
priorities for items in purgatory as well since they will be calling into
baseQueue.processReplica. This commit removes the priority reset in
replicaItem.SetProcessing(), ensuring that the enqueue time priority is retained
when replicas are popped from the heap and passed into the purgatory queue
properly. No behavior change should happen from this change.

wenyihu6 committed:
kvserver: plumb priority at enqueue for purgatory queue

Previously, replica items in the purgatory queue did not retain their enqueue
time priority. This commit ensures that the priority is preserved so it can be
passed to baseQueue.processReplica when processing items from purgatory.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152512)
<!-- Reviewable:end -->

wenyihu6 committed:
allocatorimpl: adds a priority assertion to computeAction

This commit adds an assertion to Allocator.ComputeAction to ensure that priority
is never -1 in cases where it shouldn’t be. Normally, ComputeAction returns
action.Priority(), but we sometimes adjust the priority for specific actions
like AllocatorAddVoter, AllocatorRemoveDeadVoter, and AllocatorRemoveVoter. A
priority of -1 is a special case reserved for processing logic to run even if
there’s a priority inversion. If the priority is not -1, the range may be
re-queued to be processed with the correct priority.

wenyihu6 committed:
allocatorimpl: add invariants on priority to base queue tests

This commit adds additional invariants to verify the correctness of priority
plumbing for range items in base queue tests.

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @arulajmani reviewed 18 of 18 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 4 of 4 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152512#-:-OYc5P5h6qh8eFbAOVBq:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[`-- commits` line 27 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/152512#-OYc044k5wFRWtP4WiRg:-OYc044k5wFRWtP4WiRh:b-pbdqz3):*
Agreed.
___
*[`pkg/kv/kvserver/queue.go` line 1100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152512#-OYc-QWuEi2p1KsrZ9MV:-OYc-QWv7mAMckJHXr0O:b-g4ar8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/da20ae163ff04491f3fd47305e092afe8bc4cba5/pkg/kv/kvserver/queue.go#L1100)):*
> ```Go
> // - "processing" and only in mu.replicas
> // - "purgatory" and in mu.replicas and mu.purgatory
> // 2. For every item in bq.mu.priorityQ.sl, bq.mu.purgatory, and bq.mu.replicas,
> ```

nit: maybe call this callback assertOnReplicaItem? It'll make this paragraph more readable I think.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 959 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152512#-OYbzj5XAtYChnsbqb7D:-OYbzj5XAtYChnsbqb7E:b-msi2s1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0456ce592e470721eb01dac1076bfa965411101/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L959)):*
> ```Go
> 	// priority inversion. If the priority is not -1, the range might be re-queued
> 	// to be processed with the correct priority.
> 	if priority == -1 && buildutil.CrdbTestBuild {
> ```

Should we throw in a `log.Warningf` in case it isn't a test build?


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152512)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @tbg)

___
*[`-- commits` line 27 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/152512#-OYc044k5wFRWtP4WiRg:-OYcfiRT6OqeFA1DD3cK:bryee3h):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Agreed.
</blockquote></details>

Ack, removed that TODO in the description. 
___
*[`pkg/kv/kvserver/queue.go` line 1100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/152512#-OYc-QWuEi2p1KsrZ9MV:-OYcflwm3WyNbHMnvqj0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da20ae163ff04491f3fd47305e092afe8bc4cba5/pkg/kv/kvserver/queue.go#L1100)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: maybe call this callback assertOnReplicaItem? It'll make this paragraph more readable I think.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 959 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152512#-OYbzj5XAtYChnsbqb7D:-OYcfrZF2snt_0BDu_ld:b1iebdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0456ce592e470721eb01dac1076bfa965411101/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L959)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we throw in a `log.Warningf` in case it isn't a test build?
</blockquote></details>

Good idea, done. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r= arulajmani
